<a href="https://colab.research.google.com/github/NicoD920/base_datos_primavera_2025/blob/main/EDA_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Carga de Archivos

In [154]:
import pandas as pd

url_1 = "https://raw.githubusercontent.com/NicoD920/base_datos_primavera_2025/main/archivos/olist_customers_dataset.csv"
url_2 = "https://github.com/NicoD920/base_datos_primavera_2025/raw/refs/heads/main/archivos/olist_geolocation_dataset_simple.csv"
url_3 = "https://github.com/NicoD920/base_datos_primavera_2025/raw/refs/heads/main/archivos/olist_order_items_dataset.csv"
url_4 = "https://github.com/NicoD920/base_datos_primavera_2025/raw/refs/heads/main/archivos/olist_order_payments_dataset.csv"
url_5 = "https://github.com/NicoD920/base_datos_primavera_2025//raw/refs/heads/main/archivos/olist_order_reviews_dataset.csv"
url_6 = "https://github.com/NicoD920/base_datos_primavera_2025//raw/refs/heads/main/archivos/olist_orders_dataset.csv"
url_7 = "https://github.com/NicoD920/base_datos_primavera_2025//raw/refs/heads/main/archivos/olist_products_dataset.csv"
url_8 = "https://github.com/NicoD920/base_datos_primavera_2025//raw/refs/heads/main/archivos/olist_sellers_dataset.csv"
url_9 = "https://github.com/NicoD920/base_datos_primavera_2025//raw/refs/heads/main/archivos/product_category_name_translation.csv"

df_customers = pd.read_csv(url_1)
df_geolocation = pd.read_csv(url_2)
df_items = pd.read_csv(url_3)
df_payments = pd.read_csv(url_4)
df_reviews = pd.read_csv(url_5)
df_orders = pd.read_csv(url_6)
df_products = pd.read_csv(url_7)
df_sellers = pd.read_csv(url_8)
df_category = pd.read_csv(url_9)

df_customers_pre = pd.read_csv(url_1)
df_geolocation_pre = pd.read_csv(url_2)
df_items_pre = pd.read_csv(url_3)
df_payments_pre = pd.read_csv(url_4)
df_reviews_pre = pd.read_csv(url_5)
df_orders_pre = pd.read_csv(url_6)
df_products_pre = pd.read_csv(url_7)
df_sellers_pre = pd.read_csv(url_8)
df_category_pre = pd.read_csv(url_9)

print(df_customers.head())
print(df_geolocation.head())
print(df_items.head())
print(df_payments.head())
print(df_reviews.head())
print(df_orders.head())
print(df_products.head())
print(df_sellers.head())
print(df_category.head())

                        customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP  
2                      1151              sao paulo             SP  
3                      8775        mogi das cruzes             SP  
4                     13056               campinas             SP  
   geolocation_zip_code_prefix prioritized_city geolocation_state
0                         1037        sao paulo                SP
1                    

## Nombre resultante:
df_customers

df_geolocation

df_items

df_payments

df_reviews

df_orders

df_products

df_sellers

df_translation

## Analizamos customers

In [155]:
df_customers.head()

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP  
2                      1151              sao paulo             SP  
3                      8775        mogi das cruzes             SP  
4                     13056               campinas             SP

In [156]:
df_customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   customer_id               99441 non-null  object
 1   customer_unique_id        99441 non-null  object
 2   customer_zip_code_prefix  99441 non-null  int64 
 3   customer_city             99441 non-null  object
 4   customer_state            99441 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.8+ MB


In [157]:
num_unique_customer_id = df_customers['customer_id'].nunique()
num_unique_customer_unique_id = df_customers['customer_unique_id'].nunique()

print(f"Number of unique customer_id values: {num_unique_customer_id}")
print(f"Number of unique customer_unique_id values: {num_unique_customer_unique_id}")

Number of unique customer_id values: 99441
Number of unique customer_unique_id values: 96096


In [158]:
df_customers.isnull().sum()

customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64

No se observan problemas aparentes en esta base

## Analizamos geolocation

In [159]:
df_geolocation.head()

geolocation_zip_code_prefix prioritized_city geolocation_state
0                         1037        sao paulo                SP
1                         1046        sao paulo                SP
2                         1041        sao paulo                SP
3                         1035        sao paulo                SP
4                         1012        sao paulo                SP

In [160]:
df_geolocation.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19023 entries, 0 to 19022
Data columns (total 3 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   geolocation_zip_code_prefix  19023 non-null  int64 
 1   prioritized_city             19023 non-null  object
 2   geolocation_state            19023 non-null  object
dtypes: int64(1), object(2)
memory usage: 446.0+ KB


In [213]:
df_geolocation.describe().T

count          mean           std     min  \
geolocation_zip_code_prefix  19023.0  42712.690165  30906.196252  1001.0   

                                 25%      50%      75%      max  
geolocation_zip_code_prefix  12721.5  38240.0  70663.0  99990.0

In [162]:
df_geolocation.isnull().sum()

geolocation_zip_code_prefix    0
prioritized_city               0
geolocation_state              0
dtype: int64

No se observan problemas aparentes

## Analizamos items

In [163]:
df_items.head()

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

   shipping_limit_date   price  freight_value  
0  2017-09-19 09:45:35   58.90          13.29  
1  2017-05-03 11:05:13  239.90          19.93  
2  2018-01-18 14:48:30  199.00          17.87  
3  2018-08-15 10:10:18   12.99          12.79  
4  2017-02-13 13:57:51  199.90          18.14

In [164]:
df_items.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   order_id             112650 non-null  object 
 1   order_item_id        112650 non-null  int64  
 2   product_id           112650 non-null  object 
 3   seller_id            112650 non-null  object 
 4   shipping_limit_date  112650 non-null  object 
 5   price                112650 non-null  float64
 6   freight_value        112650 non-null  float64
dtypes: float64(2), int64(1), object(4)
memory usage: 6.0+ MB


In [165]:
df_items.describe().T

count        mean         std   min    25%    50%     75%  \
order_item_id  112650.0    1.197834    0.705124  1.00   1.00   1.00    1.00   
price          112650.0  120.653739  183.633928  0.85  39.90  74.99  134.90   
freight_value  112650.0   19.990320   15.806405  0.00  13.08  16.26   21.15   

                   max  
order_item_id    21.00  
price          6735.00  
freight_value   409.68

In [166]:
df_items.isnull().sum()

order_id               0
order_item_id          0
product_id             0
seller_id              0
shipping_limit_date    0
price                  0
freight_value          0
dtype: int64

No se observan problemas aparentes

## Analizamos payments

In [167]:
df_payments.head()

order_id  payment_sequential payment_type  \
0  b81ef226f3fe1789b1e8b2acac839d17                   1  credit_card   
1  a9810da82917af2d9aefd1278f1dcfa0                   1  credit_card   
2  25e8ea4e93396b6fa0d3dd708e76c1bd                   1  credit_card   
3  ba78997921bbcdc1373bb41e913ab953                   1  credit_card   
4  42fdf880ba16b47b59251dd489d4441a                   1  credit_card   

   payment_installments  payment_value  
0                     8          99.33  
1                     1          24.39  
2                     1          65.71  
3                     8         107.78  
4                     2         128.45

In [168]:
num_unique_order_ids = df_payments['order_id'].nunique()
print(f"Number of unique order_id values in df_payments: {num_unique_order_ids}")

num_unique_payment_sequential = df_payments['payment_sequential'].nunique()
print(f"Number of unique payment_sequential values in df_payments: {num_unique_payment_sequential}")

num_unique_payment_type = df_payments['payment_type'].nunique()
print(f"Number of unique payment_type values in df_payments: {num_unique_payment_type}")

Number of unique order_id values in df_payments: 99440
Number of unique payment_sequential values in df_payments: 29
Number of unique payment_type values in df_payments: 5


In [169]:
df_payments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103886 entries, 0 to 103885
Data columns (total 5 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   order_id              103886 non-null  object 
 1   payment_sequential    103886 non-null  int64  
 2   payment_type          103886 non-null  object 
 3   payment_installments  103886 non-null  int64  
 4   payment_value         103886 non-null  float64
dtypes: float64(1), int64(2), object(2)
memory usage: 4.0+ MB


In [170]:
df_payments.describe().T

count        mean         std  min    25%    50%  \
payment_sequential    103886.0    1.092679    0.706584  1.0   1.00    1.0   
payment_installments  103886.0    2.853349    2.687051  0.0   1.00    1.0   
payment_value         103886.0  154.100380  217.494064  0.0  56.79  100.0   

                           75%       max  
payment_sequential      1.0000     29.00  
payment_installments    4.0000     24.00  
payment_value         171.8375  13664.08

In [171]:
df_payments.isnull().sum()

order_id                0
payment_sequential      0
payment_type            0
payment_installments    0
payment_value           0
dtype: int64

No se observan problemas aparentes

## Analizamos reviews

In [172]:
df_reviews.head()

review_id                          order_id  \
0  7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1  80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2  228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3  e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4  f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   

   review_score review_comment_title  \
0             4                  NaN   
1             5                  NaN   
2             5                  NaN   
3             5                  NaN   
4             5                  NaN   

                              review_comment_message review_creation_date  \
0                                                NaN  2018-01-18 00:00:00   
1                                                NaN  2018-03-10 00:00:00   
2                                                NaN  2018-02-17 00:00:00   
3              Recebi bem antes do prazo estipulado.  2017-04-21 00:00:00   
4  Parabéns lojas lannister adorei comprar pela I...  2018-03-01 00:00:00   

  review_answer_timestamp  
0     2018-01-18 21:46:59  
1     2018-03-11 03:05:13  
2     2018-02-18 14:36:24  
3     2017-04-21 22:02:06  
4     2018-03-02 10:26:53

In [173]:
df_reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99224 entries, 0 to 99223
Data columns (total 7 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   review_id                99224 non-null  object
 1   order_id                 99224 non-null  object
 2   review_score             99224 non-null  int64 
 3   review_comment_title     11568 non-null  object
 4   review_comment_message   40977 non-null  object
 5   review_creation_date     99224 non-null  object
 6   review_answer_timestamp  99224 non-null  object
dtypes: int64(1), object(6)
memory usage: 5.3+ MB


In [174]:
df_reviews.describe().T

count      mean       std  min  25%  50%  75%  max
review_score  99224.0  4.086421  1.347579  1.0  4.0  5.0  5.0  5.0

In [175]:
df_reviews.isna().sum()

review_id                      0
order_id                       0
review_score                   0
review_comment_title       87656
review_comment_message     58247
review_creation_date           0
review_answer_timestamp        0
dtype: int64

In [176]:
df_reviews.isna().sum() / df_reviews.shape[0]

review_id                  0.000000
order_id                   0.000000
review_score               0.000000
review_comment_title       0.883415
review_comment_message     0.587025
review_creation_date       0.000000
review_answer_timestamp    0.000000
dtype: float64

Eliminamos los registros con nulos dado que la cantidad es muy grande

In [177]:
df_reviews = df_reviews.drop(columns=['review_comment_title', 'review_comment_message'])
display(df_reviews.head())

review_id                          order_id  \
0  7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1  80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2  228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3  e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4  f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   

   review_score review_creation_date review_answer_timestamp  
0             4  2018-01-18 00:00:00     2018-01-18 21:46:59  
1             5  2018-03-10 00:00:00     2018-03-11 03:05:13  
2             5  2018-02-17 00:00:00     2018-02-18 14:36:24  
3             5  2017-04-21 00:00:00     2017-04-21 22:02:06  
4             5  2018-03-01 00:00:00     2018-03-02 10:26:53

## Analizamos Products

In [178]:
df_products.head()

product_id  product_category_name  \
0  1e9e8ef04dbcff4541ed26657ea517e5             perfumaria   
1  3aa071139cb16b67ca9e5dea641aaa2f                  artes   
2  96bd76ec8810374ed1b65e291975717f          esporte_lazer   
3  cef67bcfe19066a932b7673e239eb23d                  bebes   
4  9dc1a7de274444849c219cff195d0b71  utilidades_domesticas   

   product_name_lenght  product_description_lenght  product_photos_qty  \
0                 40.0                       287.0                 1.0   
1                 44.0                       276.0                 1.0   
2                 46.0                       250.0                 1.0   
3                 27.0                       261.0                 1.0   
4                 37.0                       402.0                 4.0   

   product_weight_g  product_length_cm  product_height_cm  product_width_cm  
0             225.0               16.0               10.0              14.0  
1            1000.0               30.0               18.0              20.0  
2             154.0               18.0                9.0              15.0  
3             371.0               26.0                4.0              26.0  
4             625.0               20.0               17.0              13.0

In [179]:
df_products.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32951 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32951 non-null  object 
 1   product_category_name       32341 non-null  object 
 2   product_name_lenght         32341 non-null  float64
 3   product_description_lenght  32341 non-null  float64
 4   product_photos_qty          32341 non-null  float64
 5   product_weight_g            32949 non-null  float64
 6   product_length_cm           32949 non-null  float64
 7   product_height_cm           32949 non-null  float64
 8   product_width_cm            32949 non-null  float64
dtypes: float64(7), object(2)
memory usage: 2.3+ MB


In [180]:
df_products.isnull().sum()

product_id                      0
product_category_name         610
product_name_lenght           610
product_description_lenght    610
product_photos_qty            610
product_weight_g                2
product_length_cm               2
product_height_cm               2
product_width_cm                2
dtype: int64

In [181]:
df_products.isnull().sum() / df_products.shape[0]

product_id                    0.000000
product_category_name         0.018512
product_name_lenght           0.018512
product_description_lenght    0.018512
product_photos_qty            0.018512
product_weight_g              0.000061
product_length_cm             0.000061
product_height_cm             0.000061
product_width_cm              0.000061
dtype: float64

In [182]:
# Filter rows with any null values
null_rows = df_products[df_products.isnull().any(axis=1)]

# Display the first 10 rows with null values
display(null_rows.head(10))

product_id product_category_name  \
105  a41e356c76fab66334f36de622ecbd3a                   NaN   
128  d8dee61c2034d6d075997acef1870e9b                   NaN   
145  56139431d72cd51f19eb9f7dae4d1617                   NaN   
154  46b48281eb6d663ced748f324108c733                   NaN   
197  5fb61f482620cb672f5e586bb132eae9                   NaN   
244  e10758160da97891c2fdcbc35f0f031d                   NaN   
294  39e3b9b12cd0bf8ee681bbc1c130feb5                   NaN   
299  794de06c32a626a5692ff50e4985d36f                   NaN   
347  7af3e2da474486a3519b0cba9dea8ad9                   NaN   
428  629beb8e7317703dcc5f35b5463fd20e                   NaN   

     product_name_lenght  product_description_lenght  product_photos_qty  \
105                  NaN                         NaN                 NaN   
128                  NaN                         NaN                 NaN   
145                  NaN                         NaN                 NaN   
154                  NaN                         NaN                 NaN   
197                  NaN                         NaN                 NaN   
244                  NaN                         NaN                 NaN   
294                  NaN                         NaN                 NaN   
299                  NaN                         NaN                 NaN   
347                  NaN                         NaN                 NaN   
428                  NaN                         NaN                 NaN   

     product_weight_g  product_length_cm  product_height_cm  product_width_cm  
105             650.0               17.0               14.0              12.0  
128             300.0               16.0                7.0              20.0  
145             200.0               20.0               20.0              20.0  
154           18500.0               41.0               30.0              41.0  
197             300.0               35.0                7.0              12.0  
244            2200.0               16.0                2.0              11.0  
294             300.0               16.0                7.0              11.0  
299             300.0               18.0                8.0              14.0  
347             200.0               22.0               14.0              14.0  
428            1400.0               25.0               25.0              25.0

In [183]:
print(f"Number of rows with null values in df_products: {null_rows.shape[0]}")

Number of rows with null values in df_products: 611


In [184]:
product_ids_with_nulls = df_products[df_products.isnull().any(axis=1)]['product_id'].tolist()

# print("List of product_ids with null values:")
# print(product_ids_with_nulls)

In [185]:
df_items.head()

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

   shipping_limit_date   price  freight_value  
0  2017-09-19 09:45:35   58.90          13.29  
1  2017-05-03 11:05:13  239.90          19.93  
2  2018-01-18 14:48:30  199.00          17.87  
3  2018-08-15 10:10:18   12.99          12.79  
4  2017-02-13 13:57:51  199.90          18.14

In [186]:
# Filter df_items to get orders with products from the list of null product_ids
orders_with_null_products = df_items[df_items['product_id'].isin(product_ids_with_nulls)]

# Count the number of unique order_ids in the filtered dataframe
num_orders_with_null_products = orders_with_null_products['order_id'].nunique()

print(f"Number of orders with products that have null values: {num_orders_with_null_products}")

Number of orders with products that have null values: 1452


In [187]:
# Count the total number of order_ids in the filtered dataframe (including repetitions)
num_total_orders_with_null_products = orders_with_null_products['order_id'].count()

print(f"Number of total orders with products that have null values (including repetitions): {num_total_orders_with_null_products}")

Number of total orders with products that have null values (including repetitions): 1604


In [188]:
# Calculate the total number of unique orders
total_orders = df_orders['order_id'].nunique()

# Calculate the percentage of orders with null products
percentage_orders_with_null_products = (num_orders_with_null_products / total_orders) * 100

print(f"Percentage of orders with products that have null values: {percentage_orders_with_null_products:.2f}%")

Percentage of orders with products that have null values: 1.46%


In [189]:
# Get the list of order_ids that have products with null values
order_ids_with_null_products = orders_with_null_products['order_id'].unique()

# Filter df_reviews to show reviews for these order_ids
reviews_of_orders_with_null_products = df_reviews[df_reviews['order_id'].isin(order_ids_with_null_products)]

# Display the filtered reviews
display(reviews_of_orders_with_null_products.head())

# Count the number of reviews
num_reviews_with_null_products_orders = reviews_of_orders_with_null_products.shape[0]
print(f"\nNumber of reviews for orders with products that have null values: {num_reviews_with_null_products_orders}")

review_id                          order_id  \
175  db168e3b1bb8b376934b2c5985cb3eda  4edd97a383a190e45e83c16cd6460d63   
223  3ff8285931cc9b54c9ad22ba78957f63  e6b23db78d4473c921fb9315e04b5c0f   
354  2e21bb397364fa799a4a1609fd2e6e3b  9d7f2c9f33008498700ffdc3162878fb   
440  e5a35475f7efa7f6abac5febba3fb464  7dd12269fb63d70af90a488dc5e90876   
480  004580d2657b221e45f84f4169675ecd  049ede3fc839758e81752a9dacd1f911   

     review_score review_creation_date review_answer_timestamp  
175             5  2018-03-11 00:00:00     2018-03-11 12:51:38  
223             1  2017-05-14 00:00:00     2017-05-19 10:38:27  
354             5  2017-09-01 00:00:00     2017-09-04 20:46:52  
440             5  2018-04-17 00:00:00     2018-04-19 20:58:25  
480             5  2018-08-11 00:00:00     2018-08-11 18:03:41


Number of reviews for orders with products that have null values: 1449


In [190]:
# Calculate the total number of reviews
total_reviews = df_reviews.shape[0]

# Calculate the percentage of reviews for orders with null products
percentage_reviews_with_null_products_orders = (num_reviews_with_null_products_orders / total_reviews) * 100

print(f"Percentage of reviews for orders with products that have null values: {percentage_reviews_with_null_products_orders:.2f}%")

Percentage of reviews for orders with products that have null values: 1.46%


Podemos eliminar los registros quee están nulos y solo afectarían a un 1,.46% de los datos de orders

In [191]:
# Get the list of product_ids with null values (already done in a previous cell, but redefining for clarity)
product_ids_with_nulls = df_products[df_products.isnull().any(axis=1)]['product_id'].tolist()

# Drop rows from df_products where product_id is in the list of product_ids_with_nulls
df_products = df_products[~df_products['product_id'].isin(product_ids_with_nulls)].copy()

# Get the list of order_ids associated with products that have null values (already done, but redefining for clarity)
order_ids_with_null_products = df_items[df_items['product_id'].isin(product_ids_with_nulls)]['order_id'].unique()

# Drop rows from df_items where order_id is in the list of order_ids_with_null_products
df_items = df_items[~df_items['order_id'].isin(order_ids_with_null_products)].copy()

# Drop rows from df_orders where order_id is in the list of order_ids_with_null_products
df_orders = df_orders[~df_orders['order_id'].isin(order_ids_with_null_products)].copy()

# Drop rows from df_reviews where order_id is in the list of order_ids_with_null_products
df_reviews = df_reviews[~df_reviews['order_id'].isin(order_ids_with_null_products)].copy()

print("Rows with null products and associated orders/reviews have been removed from the DataFrames.")
print("\nRemaining null values in df_products:")
print(df_products.isnull().sum())
print("\nRemaining null values in df_items:")
print(df_items.isnull().sum())
print("\nRemaining null values in df_orders:")
print(df_orders.isnull().sum())
print("\nRemaining null values in df_reviews:")
print(df_reviews.isnull().sum())

Rows with null products and associated orders/reviews have been removed from the DataFrames.

Remaining null values in df_products:
product_id                    0
product_category_name         0
product_name_lenght           0
product_description_lenght    0
product_photos_qty            0
product_weight_g              0
product_length_cm             0
product_height_cm             0
product_width_cm              0
dtype: int64

Remaining null values in df_items:
order_id               0
order_item_id          0
product_id             0
seller_id              0
shipping_limit_date    0
price                  0
freight_value          0
dtype: int64

Remaining null values in df_orders:
order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 159
order_delivered_carrier_date     1750
order_delivered_customer_date    2906
order_estimated_delivery_date       0
dt

In [192]:
df_products.isnull().sum()

product_id                    0
product_category_name         0
product_name_lenght           0
product_description_lenght    0
product_photos_qty            0
product_weight_g              0
product_length_cm             0
product_height_cm             0
product_width_cm              0
dtype: int64

In [193]:
print("############ Productos ##########")
print(df_products.info())
print("############# Reviews ###########")
print(df_reviews.info())
print("############## Items ############")
print(df_items.info())
print("############## Orders ###########")
print(df_orders.info())


############ Productos ##########
<class 'pandas.core.frame.DataFrame'>
Index: 32340 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32340 non-null  object 
 1   product_category_name       32340 non-null  object 
 2   product_name_lenght         32340 non-null  float64
 3   product_description_lenght  32340 non-null  float64
 4   product_photos_qty          32340 non-null  float64
 5   product_weight_g            32340 non-null  float64
 6   product_length_cm           32340 non-null  float64
 7   product_height_cm           32340 non-null  float64
 8   product_width_cm            32340 non-null  float64
dtypes: float64(7), object(2)
memory usage: 2.5+ MB
None
############# Reviews ###########
<class 'pandas.core.frame.DataFrame'>
Index: 97775 entries, 0 to 99223
Data columns (total 5 columns):
 #   Column                   Non-Null Cou

In [194]:
df_category.head()

product_category_name product_category_name_english
0            beleza_saude                 health_beauty
1  informatica_acessorios         computers_accessories
2              automotivo                          auto
3         cama_mesa_banho                bed_bath_table
4        moveis_decoracao               furniture_decor

In [195]:
df_category.describe().T

count unique            top freq
product_category_name            71     71   beleza_saude    1
product_category_name_english    71     71  health_beauty    1

In [196]:
df_category.isnull().sum()

product_category_name            0
product_category_name_english    0
dtype: int64

## Analizamos sellers

In [197]:
df_sellers.head()

seller_id  seller_zip_code_prefix  \
0  3442f8959a84dea7ee197c632cb2df15                   13023   
1  d1b65fc7debc3361ea86b5f14c68d2e2                   13844   
2  ce3ad9de960102d0677a81f5d0bb7b2d                   20031   
3  c0f3eea2e14555b6faeea3dd58c1b1c3                    4195   
4  51a04a8a6bdcb23deccc82b0b80742cf                   12914   

         seller_city seller_state  
0           campinas           SP  
1         mogi guacu           SP  
2     rio de janeiro           RJ  
3          sao paulo           SP  
4  braganca paulista           SP

In [198]:
df_sellers.describe().T

count          mean          std     min     25%  \
seller_zip_code_prefix  3095.0  32291.059451  32713.45383  1001.0  7093.5   

                            50%      75%      max  
seller_zip_code_prefix  14940.0  64552.5  99730.0

In [199]:
df_sellers.isnull().sum()

seller_id                 0
seller_zip_code_prefix    0
seller_city               0
seller_state              0
dtype: int64

## Analizamos category

## Analizamos orders

In [200]:
df_orders.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  
0           2017-10-18 00:00:00  
1           2018-08-13 00:00:00  
2           2018-09-04 00:00:00  
3           2017-12-15 00:00:00  
4           2018-02-26 00:00:00

In [201]:
df_orders.isnull().sum()

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 159
order_delivered_carrier_date     1750
order_delivered_customer_date    2906
order_estimated_delivery_date       0
dtype: int64

In [202]:
df_orders.isnull().sum() / df_orders.shape[0]

order_id                         0.000000
customer_id                      0.000000
order_status                     0.000000
order_purchase_timestamp         0.000000
order_approved_at                0.001623
order_delivered_carrier_date     0.017859
order_delivered_customer_date    0.029656
order_estimated_delivery_date    0.000000
dtype: float64

### Analizamos los valores faltantes según el status

In [203]:
# Identify columns with null values
cols_with_nulls = ['order_approved_at', 'order_delivered_carrier_date', 'order_delivered_customer_date']

# Group by order_status and count nulls in date columns
nulls_by_status = df_orders.groupby('order_status')[cols_with_nulls].apply(lambda x: x.isnull().sum())

display(nulls_by_status)

# Print the total number of records in df_orders
print(f"\nTotal number of records in df_orders: {df_orders.shape[0]}")

order_approved_at  order_delivered_carrier_date  \
order_status                                                    
approved                      0                             2   
canceled                    141                           538   
created                       5                             5   
delivered                    13                             2   
invoiced                      0                           306   
processing                    0                           288   
shipped                       0                             0   
unavailable                   0                           609   

              order_delivered_customer_date  
order_status                                 
approved                                  2  
canceled                                605  
created                                   5  
delivered                                 8  
invoiced                                306  
processing                              288  
shipped                                1083  
unavailable                             609


Total number of records in df_orders: 97989


In [204]:
initial_count = len(df_orders)

# Rule 1
condition_approved_null_and_status = (
    df_orders['order_approved_at'].isnull() &
    ~df_orders['order_status'].isin(['canceled', 'unavailable', 'created'])
)
removed_rule1 = condition_approved_null_and_status.sum()
df_orders = df_orders[~condition_approved_null_and_status].copy()

# Rule 2
condition_carrier_null_and_status = (
    df_orders['order_delivered_carrier_date'].isnull() &
    df_orders['order_status'].isin(['delivered', 'approved'])
)
removed_rule2 = condition_carrier_null_and_status.sum()
df_orders = df_orders[~condition_carrier_null_and_status].copy()

# Rule 3
condition_customer_null_and_status = (
    df_orders['order_delivered_customer_date'].isnull() &
    df_orders['order_status'].isin(['delivered'])
)
removed_rule3 = condition_customer_null_and_status.sum()
df_orders = df_orders[~condition_customer_null_and_status].copy()

final_count = len(df_orders)

print(f"Registros iniciales: {initial_count}")
print(f"Eliminados por Regla 1: {removed_rule1}")
print(f"Eliminados por Regla 2: {removed_rule2}")
print(f"Eliminados por Regla 3: {removed_rule3}")
print(f"Registros finales: {final_count}")
print(f"Porcentaje eliminado total: {(initial_count - final_count)/initial_count:.2%}")

print("\nValores nulos restantes:")
display(df_orders.isnull().sum())


Registros iniciales: 97989
Eliminados por Regla 1: 13
Eliminados por Regla 2: 4
Eliminados por Regla 3: 7
Registros finales: 97965
Porcentaje eliminado total: 0.02%

Valores nulos restantes:


order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 146
order_delivered_carrier_date     1746
order_delivered_customer_date    2896
order_estimated_delivery_date       0
dtype: int64

In [205]:
# --- 0) Preparación: tipos y limpieza -----------------------------------------
for d in [df_orders, df_items, df_reviews, df_payments]:
    d['order_id'] = d['order_id'].astype(str).str.strip()

# (si tienes customers/sellers/products y aparecen en items/orders, alínea tipos)
if 'customer_id' in df_orders.columns:
    df_orders['customer_id'] = df_orders['customer_id'].astype(str).str.strip()
if 'seller_id' in df_items.columns:
    df_items['seller_id'] = df_items['seller_id'].astype(str).str.strip()
if 'product_id' in df_items.columns:
    df_items['product_id'] = df_items['product_id'].astype(str).str.strip()

# --- 1) IDs remanentes y removidos --------------------------------------------
remaining_order_ids = set(df_orders['order_id'].unique())

# Si quieres auditar:
all_order_ids_before = set()
for d in [df_items, df_reviews, df_payments]:
    all_order_ids_before |= set(d['order_id'].unique())
removed_order_ids = all_order_ids_before - remaining_order_ids  # auditoría

# --- 2) Filtrar tablas dependientes por order_id -------------------------------
def filter_by_order_id(df, name):
    initial = len(df)
    df = df[df['order_id'].isin(remaining_order_ids)].copy()
    removed = initial - len(df)
    print(f"Removed {removed} rows from {name} to maintain consistency.")
    return df

df_items    = filter_by_order_id(df_items,    "df_items")
df_reviews  = filter_by_order_id(df_reviews,  "df_reviews")
df_payments = filter_by_order_id(df_payments, "df_payments")

# --- 3) (Opcional pero recomendado) Cascada hacia catálogos --------------------
# Si vas a trabajar SOLO con registros efectivamente "vigentes":
# customers que aparecen en orders remanentes
if 'customer_id' in df_orders.columns and 'df_customers' in globals():
    kept_customer_ids = set(df_orders['customer_id'].unique())
    init = len(df_customers)
    df_customers = df_customers[df_customers['customer_id'].isin(kept_customer_ids)].copy()
    print(f"Trimmed customers: {init - len(df_customers)} removed.")

# sellers y products que aparecen en items remanentes
if 'seller_id' in df_items.columns and 'df_sellers' in globals():
    kept_seller_ids = set(df_items['seller_id'].unique())
    init = len(df_sellers)
    df_sellers = df_sellers[df_sellers['seller_id'].isin(kept_seller_ids)].copy()
    print(f"Trimmed sellers: {init - len(df_sellers)} removed.")

if 'product_id' in df_items.columns and 'df_products' in globals():
    kept_product_ids = set(df_items['product_id'].unique())
    init = len(df_products)
    df_products = df_products[df_products['product_id'].isin(kept_product_ids)].copy()
    print(f"Trimmed products: {init - len(df_products)} removed.")

# --- 4) Checks de consistencia -------------------------------------------------
# Ningún DF dependiente debe contener order_id que no estén en df_orders
for name, d in [("df_items", df_items), ("df_reviews", df_reviews), ("df_payments", df_payments)]:
    assert set(d['order_id']).issubset(remaining_order_ids), f"{name} quedó con order_id huérfanos."

print("\nRemaining records:")
print(f"df_orders:   {len(df_orders)}")
print(f"df_items:    {len(df_items)}")
print(f"df_reviews:  {len(df_reviews)}")
print(f"df_payments: {len(df_payments)}")


Removed 26 rows from df_items to maintain consistency.
Removed 24 rows from df_reviews to maintain consistency.
Removed 1568 rows from df_payments to maintain consistency.
Trimmed customers: 1476 removed.
Trimmed sellers: 64 removed.
Trimmed products: 38 removed.

Remaining records:
df_orders:   97965
df_items:    110934
df_reviews:  97751
df_payments: 102318


In [206]:
status_summary = df_orders.groupby("order_status").size().sort_values(ascending=False)
print(status_summary)


order_status
delivered      95063
shipped         1083
canceled         611
unavailable      609
invoiced         306
processing       288
created            5
dtype: int64


| order\_status   | Condición para mantenerse                                                                                                                                                                               |
| --------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| **canceled**    | Se mantuvieron aunque tengan `approved_at`, `carrier_date` o `customer_date` nulos (porque es normal que no tengan esas fechas si fueron canceladas).                                                   |
| **unavailable** | Igual que canceled: se mantuvieron aunque las fechas sean nulas (porque nunca se pudieron despachar).                                                                                                   |
| **created**     | Mantienes los que tienen fechas nulas (esperado, nunca se procesaron).                                                                                                                                  |
| **delivered**   | Mantienes **solo los que tienen las tres fechas completas** (`approved_at`, `carrier_date`, `customer_date`). Los delivered con alguna fecha nula se eliminaron.                                        |
| **approved**    | Mantienes los que tienen `approved_at` (los válidos). Eliminaste los que tenían `approved_at` nulo (no tiene sentido estar en "approved" sin fecha de aprobación) y los que tenían `carrier_date` nulo. |
| **invoiced**    | Mantienes aunque tengan `carrier_date` y `customer_date` nulos (porque es lógico, todavía no se despachan).                                                                                             |
| **processing**  | Igual que invoiced, mantienes aunque tengan nulos.                                                                                                                                                      |
| **shipped**     | Mantienes aunque `customer_date` sea nulo (porque es lógico, están en tránsito).                                                                                                                        |


Sugerencia de centinelas (distintas por columna)

order_approved_at → 2090-01-01

order_delivered_carrier_date → 2090-01-02

order_delivered_customer_date → 2090-01-03

In [207]:
import pandas as pd
import numpy as np

# Asegura que las columnas sean datetime (por si vienen como string)
date_cols = ['order_approved_at', 'order_delivered_carrier_date', 'order_delivered_customer_date']
for c in date_cols:
    df_orders[c] = pd.to_datetime(df_orders[c], errors='coerce')

# Define centinelas
SENTINEL_APPROVED = pd.Timestamp('2090-01-01')
SENTINEL_CARRIER  = pd.Timestamp('2090-01-02')
SENTINEL_CUSTOMER = pd.Timestamp('2090-01-03')

# 1) approved_at nulo pero ESPERABLE (canceled, unavailable, created) -> centinela
mask_approved_expected_null = (
    df_orders['order_approved_at'].isna() &
    df_orders['order_status'].isin(['canceled', 'unavailable', 'created'])
)
df_orders['approved_was_null'] = False
df_orders.loc[mask_approved_expected_null, 'approved_was_null'] = True
df_orders.loc[mask_approved_expected_null, 'order_approved_at'] = SENTINEL_APPROVED

# 2) carrier_date nulo pero ESPERABLE (canceled, unavailable, processing, invoiced) -> centinela
mask_carrier_expected_null = (
    df_orders['order_delivered_carrier_date'].isna() &
    df_orders['order_status'].isin(['canceled', 'unavailable', 'processing', 'invoiced'])
)
df_orders['carrier_was_null'] = False
df_orders.loc[mask_carrier_expected_null, 'carrier_was_null'] = True
df_orders.loc[mask_carrier_expected_null, 'order_delivered_carrier_date'] = SENTINEL_CARRIER

# 3) customer_date nulo pero ESPERABLE (canceled, unavailable, shipped) -> centinela
mask_customer_expected_null = (
    df_orders['order_delivered_customer_date'].isna() &
    df_orders['order_status'].isin(['canceled', 'unavailable', 'shipped'])
)
df_orders['customer_was_null'] = False
df_orders.loc[mask_customer_expected_null, 'customer_was_null'] = True
df_orders.loc[mask_customer_expected_null, 'order_delivered_customer_date'] = SENTINEL_CUSTOMER


# (Opcional) columnas helper para detectar centinelas de un vistazo
df_orders['is_sentinel_approved'] = df_orders['order_approved_at'].eq(SENTINEL_APPROVED)
df_orders['is_sentinel_carrier']  = df_orders['order_delivered_carrier_date'].eq(SENTINEL_CARRIER)
df_orders['is_sentinel_customer'] = df_orders['order_delivered_customer_date'].eq(SENTINEL_CUSTOMER)


In [208]:
# Centinela ya definido antes
SENTINEL_CUSTOMER = pd.Timestamp('2090-01-03')

# Ampliar la máscara de nulo esperable en customer_date
mask_customer_expected_null = (
    df_orders['order_delivered_customer_date'].isna() &
    df_orders['order_status'].isin([
        'canceled', 'unavailable', 'shipped', 'invoiced', 'processing', 'approved', 'created'
    ])
)

# Si no existen las banderas, créalas (idempotente)
if 'customer_was_null' not in df_orders.columns:
    df_orders['customer_was_null'] = False
if 'is_sentinel_customer' not in df_orders.columns:
    df_orders['is_sentinel_customer'] = False

# Marcar y rellenar con centinela
df_orders.loc[mask_customer_expected_null, 'customer_was_null'] = True
df_orders.loc[mask_customer_expected_null, 'order_delivered_customer_date'] = SENTINEL_CUSTOMER
df_orders.loc[mask_customer_expected_null, 'is_sentinel_customer'] = True


In [209]:
df_orders.isnull().sum()

order_id                         0
customer_id                      0
order_status                     0
order_purchase_timestamp         0
order_approved_at                0
order_delivered_carrier_date     5
order_delivered_customer_date    0
order_estimated_delivery_date    0
approved_was_null                0
carrier_was_null                 0
customer_was_null                0
is_sentinel_approved             0
is_sentinel_carrier              0
is_sentinel_customer             0
dtype: int64

In [210]:
# ¿Qué estados tienen los nulos restantes?
print("Carrier nulo por status:")
print(df_orders.loc[df_orders['order_delivered_carrier_date'].isna(), 'order_status']
      .value_counts(dropna=False))

print("\nCustomer nulo por status:")
print(df_orders.loc[df_orders['order_delivered_customer_date'].isna(), 'order_status']
      .value_counts(dropna=False))


import pandas as pd

SENTINEL_CARRIER = pd.Timestamp('2090-01-02')

mask_carrier_created = (
    df_orders['order_delivered_carrier_date'].isna() &
    df_orders['order_status'].str.strip().str.lower().eq('created')
)

df_orders.loc[mask_carrier_created, 'carrier_was_null'] = True
df_orders.loc[mask_carrier_created, 'order_delivered_carrier_date'] = SENTINEL_CARRIER
df_orders.loc[mask_carrier_created, 'is_sentinel_carrier'] = True

# Comprobación final
print(df_orders['order_delivered_carrier_date'].isnull().sum())  # debería ser 0



Carrier nulo por status:
order_status
created    5
Name: count, dtype: int64

Customer nulo por status:
Series([], Name: count, dtype: int64)
0


In [211]:
df_orders.isnull().sum()

order_id                         0
customer_id                      0
order_status                     0
order_purchase_timestamp         0
order_approved_at                0
order_delivered_carrier_date     0
order_delivered_customer_date    0
order_estimated_delivery_date    0
approved_was_null                0
carrier_was_null                 0
customer_was_null                0
is_sentinel_approved             0
is_sentinel_carrier              0
is_sentinel_customer             0
dtype: int64

## RESUMEN

In [212]:
import pandas as pd
import numpy as np

# Diccionario para mapear nombres
dfs_pre = {
    "orders": df_orders_pre,
    "items": df_items_pre,
    "reviews": df_reviews_pre,
    "payments": df_payments_pre,
    "customers": df_customers_pre,
    "products": df_products_pre,
    "sellers": df_sellers_pre,
    "category": df_category_pre,
    "geolocation": df_geolocation_pre
}

dfs_post = {
    "orders": df_orders,
    "items": df_items,
    "reviews": df_reviews,
    "payments": df_payments,
    "customers": df_customers,
    "products": df_products,
    "sellers": df_sellers,
    "category": df_category,
    "geolocation": df_geolocation
}

summary_rows = []

for name, df_pre in dfs_pre.items():
    df_post = dfs_post[name]
    row = {
        "DataFrame": name,
        "Registros antes": len(df_pre),
        "Registros después": len(df_post),
        "Variación (%)": (len(df_post) - len(df_pre)) / len(df_pre) * 100,
        "Columnas": len(df_post.columns)
    }

    # Conteo total de nulos antes / después
    row["Nulos antes"] = int(df_pre.isnull().sum().sum())
    row["Nulos después"] = int(df_post.isnull().sum().sum())

    # IDs únicos relevantes
    if "order_id" in df_pre.columns:
        row["Órdenes únicas antes"] = df_pre["order_id"].nunique()
        row["Órdenes únicas después"] = df_post["order_id"].nunique()
    if "customer_id" in df_pre.columns:
        row["Clientes únicos antes"] = df_pre["customer_id"].nunique()
        row["Clientes únicos después"] = df_post["customer_id"].nunique()

    # Promedios (solo en algunos DF)
    if name == "items":
        row["Promedio items/orden antes"] = df_pre.groupby("order_id").size().mean()
        row["Promedio items/orden después"] = df_post.groupby("order_id").size().mean()
    if name == "payments":
        row["Promedio pagos/orden antes"] = df_pre.groupby("order_id").size().mean()
        row["Promedio pagos/orden después"] = df_post.groupby("order_id").size().mean()
    if name == "reviews":
        row["Promedio reviews/orden antes"] = df_pre.groupby("order_id").size().mean()
        row["Promedio reviews/orden después"] = df_post.groupby("order_id").size().mean()


    summary_rows.append(row)

df_summary = pd.DataFrame(summary_rows)

# Ordenar columnas para legibilidad
cols_order = [
    "DataFrame", "Registros antes", "Registros después", "Variación (%)",
    "Columnas", "Nulos antes", "Nulos después",
    "Órdenes únicas antes", "Órdenes únicas después",
    "Clientes únicos antes", "Clientes únicos después",
    "Promedio items/orden antes", "Promedio items/orden después",
    "Promedio pagos/orden antes", "Promedio pagos/orden después",
    "Promedio reviews/orden antes", "Promedio reviews/orden después",
]
df_summary = df_summary.reindex(columns=[c for c in cols_order if c in df_summary.columns])

# Display the DataFrame using the standard display function
display(df_summary)

DataFrame  Registros antes  Registros después  Variación (%)  Columnas  \
0       orders            99441              97965      -1.484297        14   
1        items           112650             110934      -1.523302         7   
2      reviews            99224              97751      -1.484520         5   
3     payments           103886             102318      -1.509347         5   
4    customers            99441              97965      -1.484297         5   
5     products            32951              32302      -1.969591         9   
6      sellers             3095               3031      -2.067851         4   
7     category               71                 71       0.000000         2   
8  geolocation            19023              19023       0.000000         3   

   Nulos antes  Nulos después  Órdenes únicas antes  Órdenes únicas después  \
0         4908              0               99441.0                 97965.0   
1            0              0               98666.0                 97190.0   
2       145903              0               98673.0                 97209.0   
3            0              0               99440.0                 97964.0   
4            0              0                   NaN                     NaN   
5         2448              0                   NaN                     NaN   
6            0              0                   NaN                     NaN   
7            0              0                   NaN                     NaN   
8            0              0                   NaN                     NaN   

   Clientes únicos antes  Clientes únicos después  Promedio items/orden antes  \
0                99441.0                  97965.0                         NaN   
1                    NaN                      NaN                    1.141731   
2                    NaN                      NaN                         NaN   
3                    NaN                      NaN                         NaN   
4                99441.0                  97965.0                         NaN   
5                    NaN                      NaN                         NaN   
6                    NaN                      NaN                         NaN   
7                    NaN                      NaN                         NaN   
8                    NaN                      NaN                         NaN   

   Promedio items/orden después  Promedio pagos/orden antes  \
0                           NaN                         NaN   
1                      1.141414                         NaN   
2                           NaN                         NaN   
3                           NaN                     1.04471   
4                           NaN                         NaN   
5                           NaN                         NaN   
6                           NaN                         NaN   
7                           NaN                         NaN   
8                           NaN                         NaN   

   Promedio pagos/orden después  Promedio reviews/orden antes  \
0                           NaN                           NaN   
1                           NaN                           NaN   
2                           NaN                      1.005584   
3                      1.044445                           NaN   
4                           NaN                           NaN   
5                           NaN                           NaN   
6                           NaN                           NaN   
7                           NaN                           NaN   
8                           NaN                           NaN   

   Promedio reviews/orden después  
0                             NaN  
1                             NaN  
2                        1.005576  
3                             NaN  
4                             NaN  
5                             NaN  
6                             NaN  
7                             NaN  
8     

| DataFrame       | % Variación                          | Nulos antes → después | Comentario                                                                                                                                                      |
| --------------- | ------------------------------------ | --------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| **orders**      | **-1.48 %** (↓ 1.476 ordenes aprox.) | **4.908 → 0** ✅       | Se eliminaron órdenes con fechas incoherentes y se imputaron centinelas → 0 nulos.                                                                              |
| **items**       | -1.52 %                              | 0 → 0                 | Eliminación consistente: sólo se quitaron items asociados a órdenes eliminadas. El promedio de items/orden se mantiene casi igual (**1.1417 → 1.1414**).        |
| **reviews**     | -1.48 %                              | 145.903 → 0 ✅         | Limpieza exitosa: ya no quedan reviews con valores nulos. La cantidad promedio de reviews por orden se mantiene prácticamente idéntica (**1.00558 → 1.00557**). |
| **payments**    | -1.51 %                              | 0 → 0                 | Cohesión perfecta: sólo se removieron pagos de órdenes eliminadas. Promedio pagos/orden estable (**1.0447 → 1.0444**).                                          |
| **customers**   | -1.48 %                              | 0 → 0                 | Clientes en línea con las órdenes vigentes → no hay huérfanos.                                                                                                  |
| **products**    | -1.97 %                              | 2.448 → 0 ✅           | Productos sin asociación a órdenes eliminadas fueron filtrados.                                                                                                 |
| **sellers**     | -2.06 %                              | 0 → 0                 | Solo se mantienen sellers que aún tienen items activos.                                                                                                         |
| **category**    | 0.0 %                                | 0 → 0                 | No hubo impacto: categorías permanecen intactas.                                                                                                                |
| **geolocation** | 0.0 %                                | 0 → 0                 | Sin cambios: todas las ubicaciones se mantienen.                                                                                                                |


Conclusiones de Calidad de Datos

Reducción controlada: La caída de registros está en torno a 1.5 % en casi todos los DataFrames → es consistente con el criterio aplicado (remoción de registros incoherentes).

Nulos erradicados: Todas las columnas de fechas y reviews que tenían nulos ahora están limpias o marcadas con centinelas → dataset listo para análisis sin imputaciones adicionales.

Integridad referencial mantenida:

items, reviews y payments ahora contienen solo órdenes vigentes.

customers, products, sellers alineados → no hay registros huérfanos.

Distribuciones preservadas: Los promedios (items/orden, pagos/orden, reviews/orden) son prácticamente idénticos → no distorsionaste el dataset.